Causality_analysis.py

```diff
@@ -0,0 +1,164 @@
+                                    bem=fn_bem,fname=None, meg=True, eeg=False, 
+                                    mindist=5.0,n_jobs=2, overwrite=True)
+                                    
+        fwd = mne.convert_forward_solution(fwd, surf_ori=True)
+        noise_cov = mne.read_cov(fn_cov)
+        noise_cov = mne.cov.regularize(noise_cov, epochs.info,
+                                    mag=0.05, grad=0.05, proj=True)
+        forward_meg = mne.pick_types_forward(fwd, meg=True, eeg=False)
+        inverse_operator = mne.minimum_norm.make_inverse_operator(epochs.info, 
+                                    forward_meg, noise_cov, loose=0.2, depth=0.8)
+        mne.minimum_norm.write_inverse_operator(fn_inv, inverse_operator)
+        # Compute inverse solution
+        stcs = apply_inverse_epochs(epochs, inverse_operator, lambda2, "dSPM",
+                                pick_ori=None)
+        stcs_path = MNI_dir+'/stcs/%s' %subject
+        isExists=os.path.exists(stcs_path)
+        if not isExists:
+            os.makedirs(stcs_path) 
+        s = 0
+        while s < len(stcs):
+            stc_morph = mne.morph_data(subject, 'fsaverage', stcs[s], 4, smooth=4)
+            stc_morph.save(stcs_path+'/%s_trial%s_fsaverage' %(subject, str(s)), ftype='stc')
+            s = s +1
+##################################################################
+# Causal analysis
+# 1) read the morphed individual epochs
+# 2) compute the model order
+# 3) make the significant threshold
+# 4）get the plot of effective connectivities
+##################################################################
+def causal_analysis(subject,top_c=8):
+    import numpy as np
+    import mne, os
+    subject_path = subjects_dir + 'fsaverage'
+    stcs_path = subject_path + '/stcs/%s/' %subject
+    inv_path = subject_path + '/bem/fsaverage-ico-4-src.fif'
+    src_inv = mne.read_source_spaces(inv_path, add_geom=True)
+    
+    #Get unfiltered and morphed stcs
+    stcs = []
+    i = 0
+    while i < 120:
+        fn_stc = stcs_path + '%s_trial%s_fsaverage' %(subject, str(i))
+        stc = mne.read_source_estimate(fn_stc+'-lh.stc', subject='fsaverage')
+        stcs.append(stc)
+        i = i + 1
+    #Get common labels
+    list_dirs = os.walk(subject_path + '/func_labels/common/') 
+    labels = []
+    rois = []
+    for root, dirs, files in list_dirs: 
+        for f in files: 
+            label_fname = os.path.join(root, f) 
+            label = mne.read_label(label_fname)
+            labels.append(label)
+            rois.append(f)
+    #Extract stcs in common labels
+    label_ts = mne.extract_label_time_course(stcs, labels, src_inv, mode='mean_flip',
+                                         return_generator=False)
+    
+    # Causal analysis
+    from scot.connectivity import connectivity
+    import scot.connectivity_statistics as scs
+    from scot.backend_sklearn import VAR
+    import scot.plotting as splt
```

Comment:
If there is any references to ScoT or any papers, then they can be included here.